Documentation/DesignDocs/corewcfqueue-design.md

```diff
@@ -0,0 +1,190 @@
+
+# Why are we building
+When WCF was build there were not many queing technique and there was no concept of cloud. With multiple cloud providers supporting queing technique and many queing technique widely available, we wanted to build a generic transport, so that various queuing mechanism can be supported in CoreWCF.
+
+# Class Diagram
+![CoreWCF Generic Queue](/Documentation/DesignDocs/generic_queue_class_diagram.png?raw=true)
+
+# Class Details
+#### IQueueMiddlewareBuilder and QueueMiddlewareBuilder
+- IQueueMiddlewareBuilder provides signatures for 2 important methods, ```Use``` and ```Build```. 
+```csharp
+internal interface IQueueMiddlewareBuilder 
+{ 
+   IServiceProvider QueueServices { get; set; } 
+   IDictionary<string, object> Properties { get; } 
+   IQueueMiddlewareBuilder Use(Func<QueueMessageDispatcherDelegate, QueueMessageDispatcherDelegate> middleware); 
+   QueueMessageDispatcherDelegate Build(); 
+} 
+```
+- ```Use``` adds middleware delegates to the Queue pipeline 
+- ```Build``` builds the delegate used by this application to process requests received from the queue.
+#### QueueMiddleWareBuilderExtension
+- QueueMiddleWareBuilderExtension defines ```UseMiddleware``` method which specifically search for ```InvokeAsync``` method in each middleware class and construct it as delegate to be invoked by the middleware pipeline. This is similar to the NetTcp class UseMiddlewareFramingConnectionHandshakeExtensions and the ASP.NET Core UseMiddlewareExtensions..
+```csharp
+public class QueueMiddleWareBuilderExtension
+{
+    internal const string InvokeAsyncMethodName = "InvokeAsync";
+    public static IQueueMiddlewareBuilder UseMiddleware<TMiddleware>(this IQueueMiddlewareBuilder app, params object[] args)
+    {
+         return app.UseMiddleware(typeof(TMiddleware), args);
+    }
+    
+    public static IQueueMiddlewareBuilder UseMiddleware(this IQueueMiddlewareBuilder app, Type middleware, params object[] args)
+    {
+        //TODO
+     }
+    }
+```
+#### QueueServiceWebHostBuilderExtension
+- QueueServiceWebHostBuilderExtension is the WebHostBuilder extension which stiches all the middleware pipeline and ```QueuePollingService``` which will be called during service startup for queuing mechanism to work. e:g:- 
+```csharp
+internal sealed class Startup
+{
+    public void ConfigureServices(IServiceCollection services)
+    {
+      services.AddServiceModelServices();
+      services.UseCoreWCFQueue();
+ }
+ ```
+ - It first keeps an instance of IQueueMiddlewareBuilder for chaining all the middleware components to be used
+ - Afterwards it initalizes the ```QueueHandShakeMiddleWare``` and ```QueuePollingService```.
+#### QueueHandShakeMiddleWare
+- The QueueHandShakeMiddleWare has 2 methods, ```InitQueueMiddleWare``` and ```Build```.
```

Comment:
We need to provide the class description as code. The name `InitQueueMiddleware` is very redundant. Also as it's called from the constructor, it's an implementation detail and should be a private method. There's no reason for user code to call it.  
Please specify the type of delegate which is returned.